core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt

```diff
@@ -623,3 +626,265 @@ class PathfindingTest : ApiTest() {
+        // The second intermediate stop is also moved to the end of the block (at the "DD0_9"
+        // detector's
+        // position, 37.5 meters after the second intermediate stop)
+        val recalculatedSecondIntermediateStopDistance =
+            secondIntermediateStopDistance + 37.5.meters
+
+        val parsed =
+            callPathfindingEndpoint(
+                TestTrains.VERY_LONG_FAST_TRAIN,
+                waypoints,
+                "small_infra/infra.json",
+                true,
+            )
+        checkPathfindingSuccess(
+            parsed,
+            45900.meters,
+            expectedIntermediatePathItemPosition =
+                listOf(
+                    Offset(recalculatedFirstIntermediateStopDistance),
+                    Offset(recalculatedSecondIntermediateStopDistance),
+                ),
+        )
+    }
+
+    @Test
+    fun reversedLongTrainTest() {
+        // The first intermediate stop is moved to the next block-delimiting signal, but the
+        // distance available is less than the length of the train, so it is moved to the end of the
+        // block,
+        // the 9th block of the path (1500 meters after the first intermediate stop)
+        val recalculatedFirstIntermediateStopDistance =
+            reversedFirstIntermediateStopDistance + 1500.meters
+        // The second intermediate stop is also moved to the end of the block, the 13th block of the
+        // path
+        // (370 meters after the second intermediate stop)
+        val recalculatedSecondIntermediateStopDistance =
+            reversedSecondIntermediateStopDistance + 370.meters
+
+        val parsed =
+            callPathfindingEndpoint(
+                TestTrains.VERY_LONG_FAST_TRAIN,
+                waypoints.reversed(),
+                "small_infra/infra.json",
+                true,
+            )
+        checkPathfindingSuccess(
+            parsed,
+            45900.meters,
+            expectedIntermediatePathItemPosition =
+                listOf(
+                    Offset(recalculatedFirstIntermediateStopDistance),
+                    Offset(recalculatedSecondIntermediateStopDistance),
+                ),
+        )
+    }
+
+    @Test
+    fun penultimateStopInsideDestinationBlock() {
+        var waypoints: List<List<TrackLocation>>
+        val startWaypoint = TrackLocation("TC1", Offset(500.meters))
+        val intermediateWaypoint = TrackLocation("TD0", Offset(12600.meters))
+        val endWaypoint = TrackLocation("TD0", Offset(13000.meters))
+        waypoints = listOf(listOf(startWaypoint), listOf(intermediateWaypoint), listOf(endWaypoint))
+
+        // The intermediate stop is moved to the next block-delimiting signal, but the
+        // distance available is less than the length of the train, so it is moved to the end of the
+        // block,
+        // but the end of the block is also the destination so the intermediate stop remains the
+        // same
```

Comment:
Not testing versus the end of the block here, just the destination:
```suggestion
        // distance available is more than the length of the train, so it would be moved by the length of the train (400m)
        // but this is also exactly the destination so the intermediate stop remains the same
```